### PR TITLE
Adds available method to product model and corrects products sort

### LINKED ISSUE #9634: Storefront's api v2 Products endpoint returns products with available date set in the future
## Context
When an admin creates new products in the backend it's optional to set an available date for the product. In some cases the product is not yet in stock and therefore the admin might set the available date to a future date. Spree's frontend does not show a product unless it has a set date and that date is current or in the past. However the storefront api seems to leak out products that don't yet have an available on date or have a future date.

## Expected Behavior

Products that have available date set in the future or null should not be returned from the storefront api. At least not to an unauthenticated user or non-admins. 

## Actual Behavior

Products that have available date set in the future or null are returned from the storefront api.

## Possible Fix

I’m guessing that product finder class is missing an available check in the scope.

## Steps to Reproduce

1. Deploy the lates 4.0 stable sanbox version of Spree to Heroku.
2. Log in to admin and change one product's available on date to a future date. For example Ruby on Rails Tote.
3. Call the Products endpoint (http://sandkassi4.herokuapp.com/api/v2/storefront/products).
4. Observe that the product Roby on Rails Tote is returned even though it has an available date set in the future.

## Your Environment

* Version used: 4.0
* Gemfile and Gemfile.lock as text in a Gist:
* Any relevant stack traces ("Full trace" preferred):

### PR DESCRIPTION
**Pull request by PSzyszkaSpark**
Fixes https://github.com/spree/spree/issues/9634

### CODE REVIEW

File: core/app/models/spree/product.rb [hunk 1]
```diff
@@ -435,4 +435,13 @@ def discontinue_on_must_be_later_than_available_on
       end
     end
   end
+
+  def self.available
```
Comment (damianlegawiec) [inline core/app/models/spree/product.rb]: @PSzyszkaSpark there's already an `available` scope defined here

https://github.com/spree/spree/blob/1da20381892509df9a472f450300daf32241b89d/core/app/models/concerns/spree/product_scopes.rb#L202

**Review by damianlegawiec — CHANGES_REQUESTED** [review]

**Comment by haukurhrafn** [comment]
I wonder wether it's a good idea to have  show_available as a filter. Since this is the storefront api any user can change that value and view products that have not been made available. Seems a bit “leaky” to me.

In a scenario where a store is launching an exclusive product on a future date important details might be leaked out before the set available date.

In my opinion admins are the only users that should have access to products that have not been made available. Same goes for discontinued.

File: core/app/finders/spree/products/find.rb [hunk 2]
```diff
@@ -15,6 +15,7 @@ def initialize(scope:, params:, current_currency:)
         @sort_by          = params.dig(:sort_by)
         @deleted          = params.dig(:filter, :show_deleted)
         @discontinued     = params.dig(:filter, :show_discontinued)
+        @available        = params.dig(:filter, :show_available)
```
Comment (damianlegawiec) [inline core/app/finders/spree/products/find.rb]:
```suggestion
        @not_available        = params.dig(:filter, :show_not_available)
```

File: core/app/finders/spree/products/find.rb [hunk 3]
```diff
@@ -27,14 +28,15 @@ def execute
         products = by_option_value_ids(products)
         products = include_deleted(products)
         products = include_discontinued(products)
+        products = include_available(products)
```
Comment (damianlegawiec) [inline core/app/finders/spree/products/find.rb]:
```suggestion
        products = include_not_available(products)
```

File: core/app/finders/spree/products/find.rb [hunk 4]
```diff
@@ -145,6 +147,16 @@ def include_deleted(products)
       def include_discontinued(products)
         discontinued ? products : products.not_discontinued
       end
+
+      def include_available(products)
```
Comment (damianlegawiec) [inline core/app/finders/spree/products/find.rb]:
```suggestion
      def include_not_available(products)
```

File: core/app/finders/spree/products/find.rb [hunk 5]
```diff
@@ -145,6 +147,16 @@ def include_deleted(products)
...
+        if available
```
Comment (damianlegawiec) [inline core/app/finders/spree/products/find.rb]:
```suggestion
        if not_available
```

File: core/app/models/concerns/spree/product_scopes.rb [hunk 6]
```diff
@@ -199,9 +199,9 @@ def self.not_discontinued(only_not_discontinued = true)
       end
       search_scopes << :not_discontinued
       # Can't use add_search_scope for this as it needs a default argument
-      def self.available(available_on = nil, _currency = nil)
+      def self.available(available_on = nil, only_not_discontinued = true, _currency = nil)
```
Comment (damianlegawiec) [inline core/app/models/concerns/spree/product_scopes.rb]: @PSzyszkaSpark please don't change the logic of this method

File: core/app/finders/spree/products/find.rb [hunk 7]
```diff
@@ -145,6 +147,16 @@ def include_deleted(products)
...
+          products
+        elsif discontinued
+          products.available(Time.current, '0')
+        else
+          products.available
+        end
```
Comment (damianlegawiec) [inline core/app/finders/spree/products/find.rb]:
```ruby
else
  products.available
end
```

will do fine, we don't need additional logic around discontinued

File: core/app/finders/spree/products/find.rb (hunk 3 above)
Comment (damianlegawiec) [inline core/app/finders/spree/products/find.rb]: @PSzyszkaSpark we need to change the order of scopes here

```ruby
products = include_available(products)
products = include_discontinued(products)
products = include_deleted(products)
```

**Review by damianlegawiec — CHANGES_REQUESTED** [review]

**Comment by damianlegawiec** [comment]
@haukurhrafn we don't want to change this functionality now, only fix that by default endpoints should return only available products. However, you can alter this in your project codebase - https://guides.spreecommerce.org/developer/customization/logic.html

**Review by damianlegawiec — APPROVED** [review]

**Pull request merged**